### ISSUE TITLE
Add proper support for PCF8574 port expander

### ISSUE THREAD

**Issue by dl1com** [AllYarnsAreBeautiful/ayab-firmware#211]
As discussion on Discord revealed, we are using to different types of I2C port expanders on shield and interface.
Currently, we are only supporting the MCP23008 type properly.
They are somewhat compatible in terms of their protocol, so this is not a breaking issue and may have not been an issue in the field so far, but should be fixed to avoid sporadical issues/unwanted behavior.

Support for the PCF part had been added in the past, but was removed accidentially (research done by @jonathanperret):

> Ah, I just did a bit of Git archaeology on that MCP23008/PCF8574 issue and the results are… interesting.
The issue with driving a PCF8574 with code intended for the MCP23008 was very clearly identified and solved (with a patch to the MCP23008 library) by Windell in https://github.com/AllYarnsAreBeautiful/ayab-firmware/pull/11 !
>
 >   "With this new version, the I/O signals on PCF8574-based boards are vastly improved (clean steps, no longer a forest of spikes), as seen on the logic analyzer. This may help to cut down on occasional glitches on PCF8574 based boards."
>
>This is the code that shipped in the 0.95 firmware.
Unfortunately, this customization of the MCP23008 library was overlooked as the code for the libraries was removed from the repository, in https://github.com/AllYarnsAreBeautiful/ayab-firmware/pull/27 .

**Comment by jpcornil-git** [comment]
Detection between both devices can be implemented using bit 0 of IOCON register; write 1 -> read 1 on PCF8574 but 0 on MCP23008. Based on that, the device configuration and writes to the address register can be skipped.

**Comment by jonathanperret** [comment]
> Detection between both devices can be implemented using bit 0 of IOCON register; write 1 -> read 1 on PCF8574 but 0 on MCP23008. Based on that, the device configuration and writes to the address register can be skipped.

Yes, and the code in https://github.com/AllYarnsAreBeautiful/ayab-firmware/pull/11 can be used as a reference for this approach.

**Comment by dl1com** [comment]
Is this already fixed in ayabasync? Not sure if I should flag it as v1.0.0 issue

**Comment by jpcornil-git** [comment]
No, but probably easier to implement in ayabAsync as it doesn't depend on Adafruit_MCP23008 library.
Whenever required, I can create another device for pcf8574 here https://github.com/AllYarnsAreBeautiful/ayab-firmware/tree/ayabAsync/lib/devices with the same interface as mcp23008 and instantiate one or the other after detection in the `Solenoids` [constructor](https://github.com/AllYarnsAreBeautiful/ayab-firmware/blob/d9ad53f844c1f5f554c28274945eafbae47ac7ee/lib/components/solenoids/solenoids.cpp#L3C12-L3C21).

**Comment by VIPQualityPost** [comment]
Did we end up rolling this into ayabAsync?
When we were at Open Sauce last month, we visited Sequioa Fabrica (a space that focuses on textiles) and they had ordered quite a large amount of PCF8574 boards for converting all of their machines, but ran into this issue.

**Comment by jpcornil-git** [comment]
No but it is straightforward to implement whenever required (someone will have to test the fix as I don't have an interface myself)

**Comment by VIPQualityPost** [comment]
@jpcornil-git should it target [ayabAsync branch](https://github.com/AllYarnsAreBeautiful/ayab-firmware/tree/ayabAsync) of ayab-firmware or repo on[ your account](https://github.com/jpcornil-git/ayabAsync-firmware)?

**Comment by jpcornil-git** [comment]
Agree.

@VIPQualityPost submitted already a PR on ayabAsync-firmware, see https://github.com/jpcornil-git/ayabAsync-firmware/pull/2, and I just created a pcf8574 branch with an implementation proposal to test, see https://github.com/jpcornil-git/ayabAsync-firmware/compare/v1.0...pcf8574

=> we will submit a PR into the official AYAB firmware/ayabAsync branch once the proposal tested/validated

**Comment by jpcornil-git** [comment]
@VIPQualityPost, ayabAsync branch has support for both mcp23008 and pcf8574 GPIO expanders (PR #233 merged last week) => could you give a try on your various hardware when you have time to confirm that it works as expected ?

**Comment by jpcornil-git** [comment]
Move label to 'ready - please test' ?

**Comment by VIPQualityPost** [comment]
Sorry - I am out of the country for a work trip, I can test this in a week or two when I've returned. I think it is only held back by testing now?

**Comment by jpcornil-git** [comment]
Yes, from my perspective -> why I suggested to move the label to make it clear
